client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -123,27 +125,71 @@ public ShuffleWriteClientImpl(
     this.replicaWrite = replicaWrite;
     this.replicaRead = replicaRead;
     this.replicaSkipEnabled = replicaSkipEnabled;
-    this.dataTransferPool = new ForkJoinPool(dataTranferPoolSize);
+    this.dataTransferPool = Executors.newFixedThreadPool(dataTranferPoolSize);
     this.dataCommitPoolSize = dataCommitPoolSize;
     this.unregisterThreadPoolSize = unregisterThreadPoolSize;
     this.unregisterRequestTimeSec = unregisterRequestTimeSec;
   }
 
+  @VisibleForTesting
+  public static boolean waitUntilDoneOrFail(List<Future<Boolean>> futures, boolean allowFastFail) {
+    int expected = futures.size();
+    int failed = 0;
+
+    List<Future> finished = new ArrayList<>();
+    while (true) {
+      for (Future<Boolean> future : futures) {
+        if (future.isDone() && !finished.contains(future)) {
+          finished.add(future);
+          try {
+            if (!future.get()) {
+              failed++;
+            }
+          } catch (Exception e) {
+            failed++;
+          }
+        }
+      }
+
+      if (expected == finished.size()) {
+        return failed <= 0;
+      }
+
+      if (failed > 0 && allowFastFail) {
+        futures.stream().filter(x -> !x.isDone()).forEach(x -> x.cancel(true));
+        return false;
+      }
+
+      Uninterruptibles.sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
```

Comment:
Should `10` be a configuration option?

Reply:
I also have some concern about performance.
Could we pass a timeout parameter when we call the `future.get` to solve this problem?
Could we avoid sleeping in such critical path?

Reply:
> Could we pass a timeout parameter when we call the future.get to solve this problem?
Could we avoid sleeping in such critical path?

Good idea!